Add grouping logic

semanticer opened the pull request:
## Info

### Summary
I implemented group time entries functionality as a part of our `timeEntriesLogSelector` 

### Technical considerations
I implemented this inside of the existing sequence constructor in `timeEntriesLogSelector` because half of the data processing I needed to group time entries was already there. Otherwise, I would have to "open" and group the stream later down the chain again and repeat some of the data processing again. So this seemed like the most efficient way to do this. 

I also slightly refactored `timeEntriesLogSelector` as I needed to share some of the functionality. Moreover, I was able to separate the code into two levels of detail where the main chain processing is now describing higher-level processes, whereas helper inner functions are going into more detail.

### Why is it done like this and not some other way?
I already explained some of the reasoning behind the decision to process groups in our existing `timeEntriesLogSelector` method but let me know if you have any suggestions or complaints.

### Relationships

Closes #23

---
## Repository health

### Pull Request checklist
<!-- MANDATORY: Before submiting your Pull Request, make sure that -->
<!-- all the items below are verified and the boxes are checked -->
- [x] The code does not hinder the app's performance
- [x] The code does not introduce regressions
- [x] The code in this PR is tested thoroughly enough that changing the behavior to an incorrect one breaks tests

## Reviewing

### Pointers
<!-- MANDATORY: Give pointers to help reviewers validate the changes, give a list of things that should be tested, show before/after screenshots, etc. -->

### Review template
<!-- MANDATORY: This is what reviewers should consider when reviewing your Pull request -->
<!-- Feel free to remove items that UI related if your code does not touch UI -->

<!-- General -->
- [ ] The code is thoroughly tested in well written and resilient manner
- [ ] The code is performant and leads to an optimal user experience
- [ ] The code is testable and does not cut corners for the sake of delivering
- [ ] The code considers older platforms well and will work in older devices just like it will in new ones
<!-- UI related -->
- [ ] The code considers light and dark theme
- [ ] The code considers RTL and does not mention left/right
- [ ] The code considers i18n
  - [ ] There are no hardcoded strings in the UI
  - [ ] All strings are included in every language we support OR there's an issue requesting new translations
- [ ] There are no hard-coded numbers referring to dimensions or font sizes
- [ ] The code considers UI elements that are similar and extracts their similarities into styles to avoid duplication

## Merge permissions
Anyone

heytherewill commented on timer/src/main/java/com/toggl/timer/log/domain/TimeEntriesLogSelector.kt:
```diff
@@ -13,48 +14,107 @@ fun timeEntriesLogSelector(
+
+    fun List<TimeEntry>.mapToSortedGroups(): List<List<TimeEntry>> =
+        this.groupBy(TimeEntry::similarityHashCode)
+            .map { (_, timeEntries) -> timeEntries.sortedByDescending { it.startTime } }
```
Wouldn't it be more efficient to just sort everything before?

heytherewill replied:
You didn't reply here

heytherewill commented on timer/src/main/java/com/toggl/timer/log/domain/TimeEntriesLogSelector.kt:
```diff
@@ -13,48 +14,107 @@ fun timeEntriesLogSelector(
+    }
+
+    fun List<TimeEntry>.mapToSortedGroups(): List<List<TimeEntry>> =
+        this.groupBy(TimeEntry::similarityHashCode)
```
Does the Kotlin grouping algorithm ensure order? 🤔

semanticer replied:
🤔 from groupBy docs: 

> The returned map preserves the entry iteration order of the keys produced from the original collection.

heytherewill commented on timer/src/main/java/com/toggl/timer/log/domain/TimeEntriesLogSelector.kt:
```diff
@@ -13,48 +14,107 @@ fun timeEntriesLogSelector(
+
     return timeEntries.values
         .filter { it.duration != null }
+        .filter { !it.isDeleted }
```
Ideally we shouldn't even be getting these here in the first place, but it's better safe than sorry.

Could you add it to the filter above, though, instead of having two different filters?

heytherewill commented on timer/src/main/java/com/toggl/timer/log/domain/TimeEntriesLogSelector.kt:
```diff
@@ -13,48 +14,107 @@ fun timeEntriesLogSelector(
+            sequence<TimeEntryViewModel> {
+                yieldDayHeader(groupDate, timeEntries)
+
+                if (shouldGroup)
```
I really dislike multi-line statements inside ifs without braces. It adds to the cognitive load

heytherewill replied:
I'm also not sure about these `forEachs`. They try to look function but are basically side effects (and the syntax is not the best since you are using lambdas). What's wrong with the for loop?

semanticer replied:
I guess the idea was to be consistent since I use `forEach` right above as part of the chain. But this kind of consistency probably doest really help in any way.

heytherewill commented on timer/src/main/java/com/toggl/timer/log/ui/TimeEntriesLogFragment.kt:
```diff
@@ -52,7 +52,7 @@ class TimeEntriesLogFragment : Fragment(R.layout.time_entries_log_fragment) {
 
         val curriedTimeEntriesSelector: suspend (TimeEntriesLogState) -> List<TimeEntryViewModel> = {
-            timeEntriesLogSelector(it.timeEntries, it.projects, timeService, todayString, yesterdayString)
+            timeEntriesLogSelector(it.timeEntries, it.projects, timeService, todayString, yesterdayString, false)
```
I'd rather have them always grouping, since it allows us to test everything related to the grouping logic

semanticer replied:
The issue was that grouping crashes the Adapter since we do not support group type views. I added a small dummy implementation for debugging and enabled grouping.

heytherewill commented on timer/src/main/res/layout/time_entries_group_item.xml:
```diff
@@ -0,0 +1,219 @@
+        <androidx.constraintlayout.widget.Guideline
+            android:id="@+id/top_guideline"
+            android:orientation="horizontal"
+            app:layout_constraintGuide_begin="8dp"
```
Let's extract this and reuse it both here and on the other non-grouped cell 👀

heytherewill commented on timer/src/main/res/layout/time_entries_group_item.xml:
```diff
@@ -0,0 +1,219 @@
+        android:layout_height="match_parent" />
+    <androidx.constraintlayout.widget.ConstraintLayout
+        android:id="@+id/content_view"
+        android:elevation="1dp"
```
Let's use the constants here as well.

Though, does elevation even apply here?

semanticer commented:
@heytherewill Do you have any suggestions where to put test data? TimeEntriesLogSelectorTests looks a bit crowded right now. 
As far as the layout is concerned, I added our classic counter and removed that "G..." debug string. I also implemented continue action and prepared expanding listener.
So the layout is usable for now but we should definitely do MUCH more work to share code between single and group layouts, therefore I would prefer to address this in separate issue/PR.
